### ISSUE TITLE
IRouter no longer supported in 2.2

### ISSUE THREAD

**Issue by J0nKn1ght** [aspnet/AspNetCore#7184]
In the blog post about the changes in 2.2 ([here](https://blogs.msdn.microsoft.com/webdev/2018/08/27/asp-net-core-2-2-0-preview1-endpoint-routing/)), you asked people to say what they were using IRouter for.

We have a high-profile live site, which is currently on AspNetCore 2.1, and uses IRouter to serve dynamic pages from a SQL database.  A CMS allows the sitemap to be created as a series of Page entities, and our IRouter implementation checks the incoming request, and loads the correct page from the DB.  We also support 'Live' and 'Preview' versions of the pages, which are served based on the user credentials and mode that they are viewing the site in. A user can add a new page to the sitemap, move it around the sitemap, and the live site will reflect this dynamically.  This works really well, and is combined with standard route mapping for the non-dynamic pages on the site (for example user registration, login, etc.).  The site also supports internationalisation.

However, yesterday we discovered a bug in EntityFrameworkCore v2.1 which (bizarrely) generates an exception "Cannot create a DbSet for 'X' because this type is not included in the model for the context" for a particular Linq query if the current thread's Culture is set to have a Language code of 'cs' (Czechoslovakian).  Any other language code we try (even non-existent ones) work correctly, but 'cs' doesn't for some reason.  Due to the odd nature of the circumstances which lead to the exception, it seemed likely that it was caused by a bug in EFCore, rather than our code, and upgrading to EFCore 2.2 confirmed this, as the exception no longer occurs.  We obviously also had to upgrade everything to .Net Core 2.2, and AspNetCore 2.2 in the process, though.

Because the whole site is based around the use of IRouter, I've therefore had to turn Endpoint Routing off for our project, and will have to test the whole of the site as there seem to be various errors in the routing caused by the changes in 2.2.

I'd appreciate any guidance on how to transition a site which has IRouter use at its core to use AspNetCore 2.2 and beyond, as I can't find any clear guidance on how this can be achieved.  I looked at #4221, but there doesn't seem to be much there to help with transitioning an existing site, and most of it seems to be statements of intent on how IRouter scenarios could be supported in future releases.

**Comment by J0nKn1ght** [comment]
Actually, I should have carried on investigating before posting the last comment.  The tag helpers that include 'asp-route-xx' seem to work, and the UrlHelpers that are setting the id to empty string to override the ambient values work with some controller actions, but not others, so I need to try and understand why it's inconsistent.

**Comment by pranavkm** [comment]
cc @JamesNK \ @rynowak

**Comment by J0nKn1ght** [comment]
After my rant above 😱, I've found that everything that I've tested so far still seems to work, with the exception of the one instance of Url.Action that didn't seem honour the ambient value override that I noted above.

The question about what we will be able to do to transition from IRouter still stands, though.  If there's some guidance on how this will be handled in AspNetCore 3, as well, that would be useful.
